<a href="https://colab.research.google.com/github/paraskhulbe/Data_Science_Portfolio/blob/master/A_B_testing_Results_Analytics_Workbench.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **A/B Testing Report Analytics - standard steps**


---

Analyzing the performance of variation B over A 



To be a Winner in each metric, B needs to reflect a valid and reliable better performance for the traffic in the respective metrics.

We will analyze the report in the following sequence:
*   Step 1: Sample data must correctly represent the population, if there is a sampling error, the report could not be taken into consideration further 

*  Step 2: Filtering out the metrics with the associated confidence level being higher than 95%, only reliable metrics performance will be taken for further evaluation

*  Step 3: Evaluating the B's performance for the filtered metrics


In industry, the set standards for A/B testing: 

*   Sample size (validity) : 5 days, 1-2 weeks
*   Confidence level (reliability) : >95%, 90% (in a few cases)
*   Uplift in the goal conversion rate (capability):  Value greater than the cost it takes to implement and maintenance










##Fetching the data

In [ ]:
from IPython.display import display

In [ ]:
import pandas as pd

#Connecting and Loading the Data
!wget https://www.dropbox.com/s/os1zhdtckw9qp4b/experiment-results-LIVE15OffAccessories.csv?dl=0
!wget https://www.dropbox.com/s/fbu8i6apnv07deg/experiment-results-LIVEOrganiccovertextchange.csv?dl=0

df1 = pd.read_csv("experiment-results-LIVE15OffAccessories.csv?dl=0")
df2 = pd.read_csv("experiment-results-LIVEOrganiccovertextchange.csv?dl=0")


## B -15 Off Accessories


In [ ]:
df1.head()

Start Time (UTC)       End Time (UTC)  \
0  2022-02-02 07:18:07  2022-02-07 15:25:00   
1  2022-02-02 07:18:07  2022-02-07 15:25:00   
2  2022-02-02 07:18:07  2022-02-07 15:25:00   
3  2022-02-02 07:18:07  2022-02-07 15:25:00   
4  2022-02-02 07:18:07  2022-02-07 15:25:00   

                        Variation Name  Variation ID  Is Baseline Variation  \
0  "Variation #1: 15% Off Accessories"   21129802317                  False   
1                           "Original"   21135181761                   True   
2  "Variation #1: 15% Off Accessories"   21129802317                  False   
3                           "Original"   21135181761                   True   
4  "Variation #1: 15% Off Accessories"   21129802317                  False   

                           Metric Event Name  Metric Event ID  \
0  "Total Revenue per Visitor (custom code)"      19745475478   
1  "Total Revenue per Visitor (custom code)"      19745475478   
2             "Page Visits on Checkout Page"      19741883770   
3             "Page Visits on Checkout Page"      19741883770   
4      "Unique Conversions - Thank you Page"      19743843923   

  Metric Numerator Type Metric Denominator Type Metric Winning Direction  \
0         sum - revenue                 visitor               increasing   
1         sum - revenue                 visitor               increasing   
2    unique conversions                 visitor               increasing   
3    unique conversions                 visitor               increasing   
4    unique conversions                 visitor               increasing   

   Numerator Value  Denominator Value  Metric Value  \
0       63764745.0              18806   3390.659630   
1       54701473.0              18630   2936.203596   
2           2080.0              18806      0.110603   
3           1814.0              18630      0.097370   
4            493.0              18806      0.026215   

  Improvement Status from Baseline  Improvement Value from Baseline  \
0                           better                         0.154777   
1                              NaN                              NaN   
2                           better                         0.135906   
3                              NaN                              NaN   
4                           better                         0.239559   

   Statistical Significance Is Improvement Significant  \
0                  0.282746                      False   
1                       NaN                        NaN   
2                  0.994004                       True   
3                       NaN                        NaN   
4                  0.788433                      False   

   Confidence Interval - Low  Confidence Interval - High  \
0                  -0.155606                    0.465160   
1                        NaN                         NaN   
2                   0.029302                    0.253700   
3                        NaN                         NaN   
4                  -0.056404                    0.502376   

   Samples Remaining to Significance  
0                            28260.0  
1                                NaN  
2                                0.0  
3                                NaN  
4                             6362.0

### Step 1: Validation of Sample data. It must be correctly represent the population.

In [ ]:
import numpy as np

#  Analyzing the traffic window
df1['Start Time (UTC)'] = pd.to_datetime(df1["Start Time (UTC)"])
df1['End Time (UTC)'] = pd.to_datetime(df1['End Time (UTC)'])
df1['Start Time (Weekday)'] = df1['Start Time (UTC)'].dt.day_name()
df1['End Time (Weekday)'] = df1['End Time (UTC)'].dt.day_name()
df1['Traffic Duration'] = df1['End Time (UTC)'] - df1['Start Time (UTC)'] 
df1['Traffic Duration'] = df1['Traffic Duration']/ np.timedelta64(1,'D').astype('int')
df1['Start Time (Weekday)'] = df1['Start Time (UTC)'].dt.day_name()
df1['End Time (Weekday)'] = df1['End Time (UTC)'].dt.day_name()



In [ ]:
from matplotlib import pyplot
df1_sample_size = df1[['Variation Name','Metric Event Name','Start Time (UTC)','Start Time (Weekday)','End Time (Weekday)','End Time (UTC)', 'Traffic Duration']]
df1_sample_size

Variation Name  \
0   "Variation #1: 15% Off Accessories"   
1                            "Original"   
2   "Variation #1: 15% Off Accessories"   
3                            "Original"   
4   "Variation #1: 15% Off Accessories"   
5                            "Original"   
6   "Variation #1: 15% Off Accessories"   
7                            "Original"   
8   "Variation #1: 15% Off Accessories"   
9                            "Original"   
10  "Variation #1: 15% Off Accessories"   
11                           "Original"   
12  "Variation #1: 15% Off Accessories"   
13                           "Original"   
14  "Variation #1: 15% Off Accessories"   
15                           "Original"   
16  "Variation #1: 15% Off Accessories"   
17                           "Original"   
18  "Variation #1: 15% Off Accessories"   
19                           "Original"   
20  "Variation #1: 15% Off Accessories"   
21                           "Original"   
22  "Variation #1: 15% Off Accessories"   
23                           "Original"   

                                    Metric Event Name    Start Time (UTC)  \
0           "Total Revenue per Visitor (custom code)" 2022-02-02 07:18:07   
1           "Total Revenue per Visitor (custom code)" 2022-02-02 07:18:07   
2                      "Page Visits on Checkout Page" 2022-02-02 07:18:07   
3                      "Page Visits on Checkout Page" 2022-02-02 07:18:07   
4               "Unique Conversions - Thank you Page" 2022-02-02 07:18:07   
5               "Unique Conversions - Thank you Page" 2022-02-02 07:18:07   
6                "Total Conversions - Thank you Page" 2022-02-02 07:18:07   
7                "Total Conversions - Thank you Page" 2022-02-02 07:18:07   
8                 "Clicks on Homepage - Shop Now CTA" 2022-02-02 07:18:07   
9                 "Clicks on Homepage - Shop Now CTA" 2022-02-02 07:18:07   
10  "Clicks on Homepage - Accessories Cards Shop N... 2022-02-02 07:18:07   
11  "Clicks on Homepage - Accessories Cards Shop N... 2022-02-02 07:18:07   
12                       "Visit Page: Puffy Mid Page" 2022-02-02 07:18:07   
13                       "Visit Page: Puffy Mid Page" 2022-02-02 07:18:07   
14     "Clicks on Accessories PDPs - Add to Cart CTA" 2022-02-02 07:18:07   
15     "Clicks on Accessories PDPs - Add to Cart CTA" 2022-02-02 07:18:07   
16                            "Visit Page: Puffy Lux" 2022-02-02 07:18:07   
17                            "Visit Page: Puffy Lux" 2022-02-02 07:18:07   
18                       "Visit Page: Puffy Mattress" 2022-02-02 07:18:07   
19                       "Visit Page: Puffy Mattress" 2022-02-02 07:18:07   
20                          "Visit Page: Puffy Royal" 2022-02-02 07:18:07   
21                          "Visit Page: Puffy Royal" 2022-02-02 07:18:07   
22        "Clicks on Mattress PDPs - Add to cart CTA" 2022-02-02 07:18:07   
23        "Clicks on Mattress PDPs - Add to cart CTA" 2022-02-02 07:18:07   

   Start Time (Weekday) End Time (Weekday)      End Time (UTC)  \
0             Wednesday             Monday 2022-02-07 15:25:00   
1             Wednesday             Monday 2022-02-07 15:25:00   
2             Wednesday             Monday 2022-02-07 15:25:00   
3             Wednesday             Monday 2022-02-07 15:25:00   
4             Wednesday             Monday 2022-02-07 15:25:00   
5             Wednesday             Monday 2022-02-07 15:25:00   
6             Wednesday             Monday 2022-02-07 15:25:00   
7             Wednesday             Monday 2022-02-07 15:25:00   
8             Wednesday             Monday 2022-02-07 15:25:00   
9             Wednesday             Monday 2022-02-07 15:25:00   
10            Wednesday             Monday 2022-02-07 15:25:00   
11            Wednesday             Monday 2022-02-07 15:25:00   
12            Wednesday             Monday 2022-02-07 15:25:00   
13            Wednesday             Monday 2022-02-07 15:25:00   
14            Wednesday          

**The traffic sample from Wednesday to Monday can cover the behaviour of the traffic for weekdays and weekend across the year. As per the set standards, the provided traffic window could be choosen for representing the traffic behavior across the year. The sampling complexity here comes with the time dimension. For more data, more time we will loose on testing. So we need to optimize the time window as per various complexities. We can continue with this sample.**

###Step 2: Filtering out metrics with the associated confidence level being greater than 95%

In [ ]:
significant_metrics = df1.reset_index(drop = True)[df1['Statistical Significance']>.95]['Metric Event Name'].values
print(significant_metrics)

['"Page Visits on Checkout Page"']


In [ ]:
df1_significant_metric = df1[df1['Metric Event Name'] == '"Page Visits on Checkout Page"']
df1_significant_metric['Metric Value Lower Range'] = df1_significant_metric['Metric Value'] - df1_significant_metric['Confidence Interval - Low']
df1_significant_metric['Metric Value Higher Range'] = df1_significant_metric['Metric Value'] + df1_significant_metric['Confidence Interval - High']

In [ ]:
display(df1_significant_metric[['Variation Name','Metric Event Name','Metric Value','Metric Value Lower Range','Metric Value Higher Range','Statistical Significance']])

Variation Name               Metric Event Name  \
2  "Variation #1: 15% Off Accessories"  "Page Visits on Checkout Page"   
3                           "Original"  "Page Visits on Checkout Page"   

   Metric Value  Metric Value Lower Range  Metric Value Higher Range  \
2      0.110603                  0.081301                   0.364303   
3      0.097370                       NaN                        NaN   

   Statistical Significance  
2                  0.994004  
3                       NaN

In [ ]:
uplift = (df1_significant_metric['Metric Value'].values[0] - df1_significant_metric['Metric Value'].values[1])/ df1_significant_metric['Metric Value'].values[1] * 100

display("Uplift(%):"+  str(round(uplift,1)))

'Uplift(%):13.6'

In [ ]:

confidence_interval = [round((df1_significant_metric['Metric Value Lower Range'].values[0] - df1_significant_metric['Metric Value'].values[1])/df1_significant_metric['Metric Value'].values[1] * 100),round((df1_significant_metric['Metric Value Higher Range'].values[0] - df1_significant_metric['Metric Value'].values[1])/df1_significant_metric['Metric Value'].values[1] * 100)]
display("Confidence Interval (%): "+str(confidence_interval))

'Confidence Interval (%): [-17, 274]'

With 99% confidence level, there is a uplift of 13.6% in the performance of the metric - "Page Visits on Checkout page". Winner!

## B - LIVE Organic cover text change

In [ ]:
df2.head()

Start Time (UTC)       End Time (UTC)  \
0  2022-05-01 09:33:38  2022-05-04 14:20:00   
1  2022-05-01 09:33:38  2022-05-04 14:20:00   
2  2022-05-01 09:33:38  2022-05-04 14:20:00   
3  2022-05-01 09:33:38  2022-05-04 14:20:00   
4  2022-05-01 09:33:38  2022-05-04 14:20:00   

                              Variation Name  Variation ID  \
0                                 "Original"   21506150440   
1  "Variation #1: Organic cover text change"   21507701126   
2                                 "Original"   21506150440   
3  "Variation #1: Organic cover text change"   21507701126   
4                                 "Original"   21506150440   

   Is Baseline Variation                          Metric Event Name  \
0                   True  "Total Revenue per Visitor (custom code)"   
1                  False  "Total Revenue per Visitor (custom code)"   
2                   True             "Page Visits on Checkout Page"   
3                  False             "Page Visits on Checkout Page"   
4                   True      "Unique Conversions - Thank you Page"   

   Metric Event ID Metric Numerator Type Metric Denominator Type  \
0      19745475478         sum - revenue                 visitor   
1      19745475478         sum - revenue                 visitor   
2      19741883770    unique conversions                 visitor   
3      19741883770    unique conversions                 visitor   
4      19743843923    unique conversions                 visitor   

  Metric Winning Direction  Numerator Value  Denominator Value  Metric Value  \
0               increasing       24313025.0               5392   4509.092174   
1               increasing       23873018.0               5037   4739.531070   
2               increasing            818.0               5392      0.151706   
3               increasing            770.0               5037      0.152869   
4               increasing            182.0               5392      0.033754   

  Improvement Status from Baseline  Improvement Value from Baseline  \
0                              NaN                              NaN   
1                           better                         0.051105   
2                              NaN                              NaN   
3                           better                         0.007663   
4                              NaN                              NaN   

   Statistical Significance Is Improvement Significant  \
0                       NaN                        NaN   
1                  0.679274                      False   
2                       NaN                        NaN   
3                  0.000000                      False   
4                       NaN                        NaN   

   Confidence Interval - Low  Confidence Interval - High  \
0                        NaN                         NaN   
1                  -0.476495                    0.232841   
2                        NaN                         NaN   
3                  -0.150317                    0.169464   
4                        NaN                         NaN   

   Samples Remaining to Significance  
0                                NaN  
1                            32689.0  
2                                NaN  
3                            50966.0  
4                                NaN

### Step 1: Validation of Sample data.

In [ ]:
#  Analyzing the traffic window
df2['Start Time (UTC)'] = pd.to_datetime(df2["Start Time (UTC)"])
df2['End Time (UTC)'] = pd.to_datetime(df2['End Time (UTC)'])
df2['Start Time (Weekday)'] = df2['Start Time (UTC)'].dt.day_name()
df2['End Time (Weekday)'] = df2['End Time (UTC)'].dt.day_name()
df2['Traffic Duration'] = df2['End Time (UTC)'] - df2['Start Time (UTC)'] 
df2['Traffic Duration'] = df2['Traffic Duration']/ np.timedelta64(1,'D').astype('int')
df2['Start Time (Weekday)'] = df2['Start Time (UTC)'].dt.day_name()
df2['End Time (Weekday)'] = df2['End Time (UTC)'].dt.day_name()

In [ ]:
df2_sample_size = df2[['Variation Name','Metric Event Name','Start Time (UTC)','Start Time (Weekday)','End Time (Weekday)','End Time (UTC)', 'Traffic Duration']]
df2_sample_size

Variation Name  \
0                                  "Original"   
1   "Variation #1: Organic cover text change"   
2                                  "Original"   
3   "Variation #1: Organic cover text change"   
4                                  "Original"   
5   "Variation #1: Organic cover text change"   
6                                  "Original"   
7   "Variation #1: Organic cover text change"   
8                                  "Original"   
9   "Variation #1: Organic cover text change"   
10                                 "Original"   
11  "Variation #1: Organic cover text change"   
12                                 "Original"   
13  "Variation #1: Organic cover text change"   
14                                 "Original"   
15  "Variation #1: Organic cover text change"   
16                                 "Original"   
17  "Variation #1: Organic cover text change"   
18                                 "Original"   
19  "Variation #1: Organic cover text change"   
20                                 "Original"   
21  "Variation #1: Organic cover text change"   
22                                 "Original"   
23  "Variation #1: Organic cover text change"   
24                                 "Original"   
25  "Variation #1: Organic cover text change"   
26                                 "Original"   
27  "Variation #1: Organic cover text change"   
28                                 "Original"   
29  "Variation #1: Organic cover text change"   

                                    Metric Event Name    Start Time (UTC)  \
0           "Total Revenue per Visitor (custom code)" 2022-05-01 09:33:38   
1           "Total Revenue per Visitor (custom code)" 2022-05-01 09:33:38   
2                      "Page Visits on Checkout Page" 2022-05-01 09:33:38   
3                      "Page Visits on Checkout Page" 2022-05-01 09:33:38   
4               "Unique Conversions - Thank you Page" 2022-05-01 09:33:38   
5               "Unique Conversions - Thank you Page" 2022-05-01 09:33:38   
6                "Total Conversions - Thank you Page" 2022-05-01 09:33:38   
7                "Total Conversions - Thank you Page" 2022-05-01 09:33:38   
8                               "Average Order Value" 2022-05-01 09:33:38   
9                               "Average Order Value" 2022-05-01 09:33:38   
10                "Clicks on Homepage - Shop Now CTA" 2022-05-01 09:33:38   
11                "Clicks on Homepage - Shop Now CTA" 2022-05-01 09:33:38   
12                    "Clicks on Homepage - Quiz CTA" 2022-05-01 09:33:38   
13                    "Clicks on Homepage - Quiz CTA" 2022-05-01 09:33:38   
14     "Clicks on Homepage - Mattress Cards CTA/Link" 2022-05-01 09:33:38   
15     "Clicks on Homepage - Mattress Cards CTA/Link" 2022-05-01 09:33:38   
16  "Clicks on Homepage - Mattress Cards Compare CTA" 2022-05-01 09:33:38   
17  "Clicks on Homepage - Mattress Cards Compare CTA" 2022-05-01 09:33:38   
18        "Clicks on Mattress PDPs - Add to Cart CTA" 2022-05-01 09:33:38   
19        "Clicks on Mattress PDPs - Add to Cart CTA" 2022-05-01 09:33:38   
20                           "Page Visits on Midpage" 2022-05-01 09:33:38   
21                           "Page Visits on Midpage" 2022-05-01 09:33:38   
22  "Clicks on Midpage - Images/Text Links to Matt... 2022-05-01 09:33:38   
23  "Clicks on Midpage - Images/Text Links to Matt... 2022-05-01 09:33:38   
24                "Page Visits on Puffy Lux Mattress" 2022-05-01 09:33:38   
25                "Page Visits on Puffy Lux Mattress" 2022-05-01 09:33:38   
26              "Page Visits on Puffy Royal Mattress" 2022-05-01 09:33:38   
27              "Page Visits on Puffy Royal Mattress" 2022-05-01 09:33:38   
28                    "Page Visits on Puffy Mattress" 2022-05-01 09:33:38   
29                    "Page Visits on Puffy Mattress" 2022-05-01 09:33:38   

   Start Time (Weekday) End Time (Weekday)      End Time (UTC)  \
0                Sunday          Wednesday 2022-05-04 14:

The data is of only 3 days traffic. **As per the set standards, the sample size is not big enough to represent the population in a valid way. Hence, we have insuffiecient data for evaluating the performance.**

## Report

In [ ]:
#@title
import plotly.graph_objects as go

fig = go.Figure(data=[go.Table(
  header=dict(
    values=['','<b>Unique Conversions <br>/ visitors<br></b>','<b>Conversion Rate</b>','<b>Improvement</b>','<b>Confidence Interval </b>','<b>Statistical Significance</b>'],
    line_color='white',
    fill_color ='white',
    align=['left','center'],
    font=dict(color='grey', size=12)
  ),
  cells=dict(
    values=[
      ['Original', '<b>Variation</b><br>#1: 15% Off Accessories'],
      [str(int(df1_significant_metric['Numerator Value'].values[1]))+str(" / ")+ str(int(df1_significant_metric['Denominator Value'].values[1])),str(int(df1_significant_metric['Numerator Value'].values[0]))+str(" / ")+ str(int(df1_significant_metric['Denominator Value'].values[0]))],
      [str(round(df1_significant_metric['Metric Value'].values[1]*100,2)) +str("%"),str(round(df1_significant_metric['Metric Value'].values[0]*100,2))+str("%")],
      ['--','<b>'+str("+")+ str(round(uplift,1))+str("%")+'</b>'],
      ['--', '  [---------]'+'<br>'+str(confidence_interval[0])+'%'+'    +'+str(confidence_interval[1])+str("%")],
      ['--<br>baseline', '>99%<br>winner']],
    line_color='white',
    align = ['left', 'center'],
    font = dict(color = [['grey','green']], size = 11)
    ))
])
fig.update_layout(title_text="<b>Page Visits on Checkout Page</b><br>Increase in unique conversions for \"Page Visits on Checkout Page\" event")

fig.show()

 